File: engine/engine.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Computational Engine                                                  *
 *                                                                        *
 *  Copyright (c) 1999-2016, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,       *
 *  MA 02110-1301, USA.                                                   *
 *                                                                        *
 **************************************************************************/

#include <regina-config.h>

#include "engine.h"
#include "utilities/intutils.h"

namespace regina {

const char* versionString() {
    return PACKAGE_VERSION;
}

int versionMajor() {
    return PACKAGE_VERSION_MAJOR;
}

int versionMinor() {
    return PACKAGE_VERSION_MINOR;
}

bool versionUsesUTF8(const char* version) {
    // No version is fine.
    // Also any version that does not begin with 0-4 is fine.
    if (version[0] < '0' || version[0] > '4')
        return true;

    if (version[0] < '4') {
        // Deal with 0.x .. 3.x.
        // Unless this is the beginning of a larger number,
        // this is a bad version.
        return (version[0] != '0' &&
            version[1] >= '0' && version[1] <= '9');
    } else {
        // The 4.x case is a bit messier to deal with.

        // If it's the beginning of a larger number, we're fine.
        if (version[1] >= '0' && version[1] <= '9')
            return true;

        // Definitely begins with 4 and 4 alone.
        // The only way to save ourselves is to have 4.x for x >= 5.
        if (version[1] != '.')
            return false;

        // We definitely begin with "4.".
        // The only good possibilities now are to begin with:
        // - 4.[number larger than 4]
        if (version[2] == '0')
            return false;
        else if (version[2] >= '1' && version[2] <= '4')
            return (version[3] >= '0' && version[3] <= '9');
        else if (version[2] >= '5' && version[2] <= '9')
            return true;
        else
            return false;
    }
}

const char* versionSnapPy() {
    return SNAPPY_VERSION;
}

const char* versionSnapPea() {
    return SNAPPY_VERSION;
}

bool hasInt128() {
#ifdef INT128_AVAILABLE
    return true;
#else
    return false;
#endif
}

int testEngine(int value) {
    return value;
}

} // namespace regina

File: javaui/normal/engine/Engine.java

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Java User Interface                                                   *
 *                                                                        *
 *  Copyright (c) 1999-2002, Ben Burton                                   *
 *  For further details contact Ben Burton (benb@acm.org).                *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,        *
 *  MA 02111-1307, USA.                                                   *
 *                                                                        *
 **************************************************************************/

/* end stub */

package normal.engine;

import normal.engine.algebra.*;
import normal.engine.angle.*;
import normal.engine.file.*;
import normal.engine.maths.*;
import normal.engine.packet.*;
import normal.engine.progress.*;
import normal.engine.split.*;
import normal.engine.subcomplex.*;
import normal.engine.surfaces.*;
import normal.engine.triangulation.*;
import normal.engine.utilities.*;

/**
 * Represents the link between Java and the calculation engine,
 * which may be external.
 * See the Java notes page in the engine documentation for further
 * details.
 * <p>
 * Specific engine implementations can be found in the package
 * <tt>normal.engine.implementation</tt>.
 */
public interface Engine {
    /**
     * Returns the style of engine represented by this engine object.
     * This will be one of the predefined engine style constants in
     * class <tt>normal.Shell</tt>.
     *
     * @return the style of engine.
     */
    int style();
    /**
     * Returns a short description of the style of engine represented by
     * this engine object.
     *
     * @return a description of the style of engine.
     */
    String styleDescription();

    NAbelianGroup newNAbelianGroup();
    NAbelianGroup newNAbelianGroup(NAbelianGroup cloneMe);
    NAngleStructureList newNAngleStructureList(NTriangulation owner);
    NContainer newNContainer();
    NFile newNFile();
    NGroupExpression newNGroupExpression();
    NGroupExpression newNGroupExpression(NGroupExpression cloneMe);
    NGroupPresentation newNGroupPresentation();
    NGroupPresentation newNGroupPresentation(NGroupPresentation cloneMe);
    NLayeredChain newNLayeredChain(NTetrahedron tet, NPerm vertexRoles);
    NLayeredChain newNLayeredChain(NLayeredChain cloneMe);
    NLensSpace newNLensSpace(long p, long q);
    NLensSpace newNLensSpace(NLensSpace cloneMe);
    NMatrixInt newNMatrixInt(int rows, int columns);
    NMatrixInt newNMatrixInt(NMatrixInt cloneMe);
    NNormalSurfaceList newNNormalSurfaceList(NTriangulation owner, int flavour,
        boolean isEmbeddedOnly);
    NNormalSurfaceList newNNormalSurfaceList(NTriangulation owner, int flavour);
    NProgressManager newNProgressManager();
    NSFS newNSFS();
    NSFS newNSFS(long orbitGenus, boolean orbitOrientable);
    NSFS newNSFS(long orbitGenus, boolean orbitOrientable, long orbitPunctures);
    NSFS newNSFS(NSFS cloneMe);
    NScript newNScript();
    NSignature newNSignature(NSignature cloneMe);
    NSurfaceFilter newNSurfaceFilter();
    NSurfaceFilter newNSurfaceFilter(NSurfaceFilter cloneMe);
    NSurfaceFilterCombination newNSurfaceFilterCombination();
    NSurfaceFilterCombination newNSurfaceFilterCombination(
        NSurfaceFilterCombination cloneMe);
    NSurfaceFilterProperties newNSurfaceFilterProperties();
    NSurfaceFilterProperties newNSurfaceFilterProperties(
        NSurfaceFilterProperties cloneMe);
    NSurfaceSubset newNSurfaceSubset(NSurfaceSet set, NSurfaceFilter filter);
    NTetrahedron newNTetrahedron();
    NTetrahedron newNTetrahedron(String desc);
    NText newNText();
    NText newNText(String text);
    NTriangulation newNTriangulation();
    NTriangulation newNTriangulation(NTriangulation cloneMe);

    NTriangulation enterTextTriangulation();
    long formCensus(NPacket parent, int nTetrahedra,
        NBoolSet finiteness, NBoolSet orientability, NBoolSet boundary);
    long formCensus(NPacket parent, int nTetrahedra,
        NBoolSet finiteness, NBoolSet orientability, NBoolSet boundary,
        int nBdryFaces);
    long formCensus(NPacket parent, int nTetrahedra,
        NBoolSet finiteness, NBoolSet orientability, NBoolSet boundary,
        int nBdryFaces, NProgressManager manager);
    NPillowTwoSphere formsPillowTwoSphere(NFace face1, NFace face2);
    NSnappedTwoSphere formsSnappedTwoSphere(NSnappedBall ball1,
        NSnappedBall ball2);
    NSnappedTwoSphere formsSnappedTwoSphere(NTetrahedron tet1,
        NTetrahedron tet2);
    int getVersionMajor();
    int getVersionMinor();
    String getVersionString();
    NFileInfo identifyFileInfo(String idPathname);
    NAugTriSolidTorus isAugTriSolidTorus(NComponent comp);
    NLayeredLensSpace isLayeredLensSpace(NComponent comp);
    NLayeredLoop isLayeredLoop(NComponent comp);
    NLayeredSolidTorus isLayeredSolidTorusBase(NTetrahedron tet);
    NSnappedBall isSnappedBall(NTetrahedron tet);
    NSpiralSolidTorus isSpiralSolidTorus(NTetrahedron tet, NPerm roles);
    NTriSolidTorus isTriSolidTorus(NTetrahedron tet, NPerm vertexRoles);
    NMatrixInt makeMatchingEquations(NTriangulation triangulation, int flavour);
    NSignature parseSignature(String sig);
    NPacket readFileMagic(String fileName);
    NPacket readFromFile(String fileName);
    NPacket readXMLFile(String fileName);
    NTriangulation readSnapPea(String file);
    void smithNormalForm(NMatrixInt matrix);
    int testEngine(int value);
    boolean writeSnapPea(String fileName, NTriangulation tri);
    boolean writeToFile(String fileName, NPacket packet);
    boolean writeXMLFile(String fileName, NPacket subtree);
    boolean writeXMLFile(String fileName, NPacket subtree, boolean compressed);
}

